docs/rfc/rfc-2.adoc

```diff
@@ -0,0 +1,811 @@
+  `getPublicKey` view method.
+- The deposit redemption script, straightforwardly computable from the keep's
+  public key by compressing the public key; this is briefly covered in the
+  implementation <<output-script>> section.
+- One Bitcoin address per signer, specified on the client in a way detailed
+  in the section on <<recovery-address>>.
+- The keep's signature on the sighash produced by the previous items,
+  computed via the threshold ECDSA protocol in a way detailed in the section on
+  the <<signing-protocol>>.
+- F, the desired fee, which is a function of the going feerate on the Bitcoin
+  network and the number of bytes in the final transaction.
+
+The desired fee is the most problematic one to procure under the goal of not
+strictly requiring a Bitcoin node connection, as generally a Bitcoin node
+provides estimates of the feerates that will achieve transaction confirmation
+within a certain amount of time. A short-term solution is proposed in the
+section on <<bitcoin-fee>>.
+
+[[recovery-address]]
+===== Specifying an Operator's Bitcoin Recovery Address
+
+To specify a given operator's Bitcoin Recovery Address, this proposal
+suggests adding a single key to the `Extensions.TBTC.BTCRefunds` config
+entry, `BeneficiaryAddress`. The value for this key should be one of:
+
+- A simple BTC address.
+- An xpub, ypub, or zpub that allows the client to derive new addresses for
+  each refund.
+
+The challenge with a *pub is that, to avoid reuse, it needs to do one of two
+things:
+
+- Be able to check, on-chain, whether a given address has been used; or,
+- Have a *pub that is only meant for use by the client, and have the client
+  track the latest address generated (as an index starting at 0).
+
+To avoid _strictly requiring_ a connection to a Bitcoin node, this RFC
+proposes tracking the latest address generated, and allowing an additional
+config key, `TargetDerivationStartIndex`, that sets a starting point for such
+derivations. The client should write the updated index to the `current/`
+directory that tracks active key shares, to an encrypted file named
+`current-btc-refund-derivation-index`. The operator should then seek to
+obtain from the staker an extended public key that is only used for this
+purpose. In cases where in-memory/on-disk derivation index and config file index
+both exist, the highest index should be used.
+
+If an optional Bitcoin node connection is available (see <<electrs>>), the
+client should additionally check a derived address for a balance before
+publishing. To do this, a GET request to `/address/<address>` should be issued,
+and the JSON response inspected for a `tx_count` of 0. When such a connection is
+available, if the client finds an address is already used, it should increment
+the derivation index and check again. If the GET request fails, the index should
+be assumed to be valid, as if the Bitcoin node connection were not available.
```

Comment:
checking for understanding:

If the signer wants to use a *pub to generate wallet addresses, the way those generators work is that they have some sort of index (in the same way that a psuedo random number generator has a seed). Since the user would supply no other information besides the generator, we can either:

* generates addresses and then check to see if they've been used before
* not care
* start at 0, assume they've never been used, and maintain internal state

(and other schemes that are more complicated)

We can't just not-care, because if they're bothering to send us a *pub instead of an address, they probably have a large preference toward remaining anonymous, and sending the same wallet multiple transactions over time creates patterns and bleeds information.

We can't check to see if they've been used before without actually connecting to a blockchain, which we're trying to avoid.

We also don't have a great way to manage state other than writing to files so we do it in-memory and read/write to an encrypted file

Reply:
Yep, you nailed it. And we provide the config option to skip “start at 0”.